core/src/main/java/com/sequenceiq/cloudbreak/job/stackpatcher/ExistingStackPatcherJobInitializer.java

```diff
@@ -27,15 +29,34 @@
     @Inject
     private ExistingStackPatcherJobService jobService;
 
+    @Inject
+    private StackPatchService stackPatchService;
+
     @Override
     public void initJobs() {
         Set<StackPatchType> enabledStackPatchTypes = getEnabledStackPatchTypes();
         LOGGER.info("Existing stack patch types enabled: {}", enabledStackPatchTypes);
 
         List<JobResource> stacks = getAliveJobResources();
+        Set<Long> stackIds = stacks.stream()
+                .map(JobResource::getLocalId)
+                .map(Long::valueOf)
+                .collect(Collectors.toSet());
         enabledStackPatchTypes.forEach(stackPatchType -> {
             LOGGER.info("Scheduling stack patcher jobs for {}", stackPatchType);
-            stacks.forEach(stack -> jobService.schedule(new ExistingStackPatcherJobAdapter(stack, stackPatchType)));
+            Map<Long, List<StackPatch>> stackPatchesByStack = stackPatchService.findAllByTypeForStackIds(stackPatchType, stackIds).stream()
+                    .collect(Collectors.groupingBy(StackPatch::getStackId));
+
+            for (JobResource stack : stacks) {
+                Long stackId = Long.valueOf(stack.getLocalId());
+                List<StackPatch> stackPatches = stackPatchesByStack.get(stackId);
+                boolean stackPatchIsFinalized = stackPatches != null && stackPatches.stream().anyMatch(stackPatch -> stackPatch.getStatus().isFinal());
```

Comment:
Minor: you should not care about null stackPatches if you are iterating over the entryset of stackPatchesByStack (instead if iterating over stacks and lookup the list of stack patches from the map)

Reply:
Ok, I see.